lua/grapple/ui/tags.lua

```diff
@@ -1,110 +1,196 @@
+            return nil
+        end
+
+        ---@type Grapple.PartialTag
+        local partial_tag = {
+            file_path = scope_path .. "/" .. relative_path,
+            key = tonumber(key) or key,
+        }
+
+        return partial_tag
     end
-    local parsed_key = string.sub(line, start + 1, _end - 1)
-    return tonumber(parsed_key) or parsed_key
 end
 
----@param scope Grapple.Scope
----@param tags Grapple.TagTable
----@param _popup Grapple.Popup
-local function resolve(scope, tags, _popup)
-    local lines = vim.api.nvim_buf_get_lines(_popup.buffer, 0, -1, false)
-    local remaining_keys = {}
-    for _, line in ipairs(lines) do
-        local tag_key = parse(line)
-        if tag_key ~= nil then
-            remaining_keys[tag_key] = true
+---@param scope_ Grapple.Scope
+---@param popup_ Grapple.Popup
+---@param parser Grapple.Parser<Grapple.PartialTag>
+local function resolve(scope_, popup_, parser)
+    ---@type string[]
+    local lines = vim.api.nvim_buf_get_lines(popup_.buffer, 0, -1, false)
+
+    ---@type Grapple.PartialTag[]
+    local partial_tags = vim.tbl_map(parser, lines)
+
+    -- Use the line number as the index for numbered tags
+    local index = 1
+    for i = 1, #partial_tags do
+        if type(partial_tags[i].key) == "number" then
+            partial_tags[i].key = index
+            index = index + 1
         end
     end
-    for key, tag in pairs(tags) do
-        if not remaining_keys[key] then
-            _tags.untag(scope, { file_path = tag.file_path })
+
+    ---@type table<string, boolean>
+    local remaining_tags = {}
+
+    ---@type Grapple.PartialTag[]
+    local modified_tags = {}
+
+    -- Determine which tags have been modified and which were deleted
+    for _, partial_tag in ipairs(partial_tags) do
+        local key = tags.key(scope_, { file_path = partial_tag.file_path })
+        if key ~= nil then
+            if partial_tag.key ~= key then
+                table.insert(modified_tags, partial_tag)
+            end
+            remaining_tags[key] = true
+        end
+    end
+
+    -- Delete tags that do not exist anymore
+    for _, key in ipairs(tags.keys(scope_)) do
+        if not remaining_tags[key] then
+            tags.untag(scope_, { key = key })
         end
     end
+
+    -- Update tags that now have a different key
+    for _, partial_tag in ipairs(modified_tags) do
+        tags.tag(scope_, { file_path = partial_tag.file_path, key = partial_tag.key })
+    end
+
+    -- Fill any "holes" that were made from deletion and updating
+    tags.reorder(scope_)
 end
 
----@param scope Grapple.Scope
----@param tags Grapple.TagTable
----@param _popup Grapple.Popup
----@return function
-local function action_close(scope, tags, _popup)
+---@param scope_ Grapple.Scope
+---@param popup_ Grapple.Popup
+---@param parser Grapple.Parser<Grapple.PartialTag>
+---@param serializer_ Grapple.Serializer<Grapple.PartialTag>
+local function action_update(scope_, popup_, parser, serializer_)
```

Comment:
Not sure if this feature will be kept

Reply:
I don't think it should, now that the `resolve` is taking care of changing tag keys also

Reply:
Sounds good. I'll remove it for now. Although, I think this could be reintroduced with an additional constraint where the buffer's undo history is cleared after an update using something like [`:h undolevels`](https://vimhelp.org/options.txt.html#%27undolevels%27)